### PR TITLE
HBase terraform 支持

### PR DESCRIPTION
**Pull request by joyken**
data_source_alicloud_hbase_instances.go
resource_alicloud_hbase_instance.go
examples/hbase
website/docs/r/
website/docs/d/
hbase产品支持
Acceptance Test：
--- PASS: TestAccAlicloudHBaseInstancesDataSourceNewInstance (906.81s)
PASS
--- PASS: TestAccAlicloudHBaseInstanceClassic (887.05s)
PASS
cloud_efficiency 100 cn-shenzhen cn-shenzhen-b open open}
--- PASS: TestAccAlicloudHBaseInstanceVpc (836.51s)
PASS
--- PASS: TestAccAlicloudHBaseInstanceMultiInstance (899.66s)
PASS

### CODE REVIEW

**Comment by Lexsss** [comment]
1. you can run `sudo make fmt && sudo make test && sudo make vendor-status && sudo make vet` to modify your codes and pass the CI check.
2. make sure your pull request has 1 commit only.

File: alicloud/service_alicloud_hbase.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+		ColdStorageStatus:  response.ColdStorageStatus,
+		BackupStatus:       response.BackupStatus,
+	}
+	return instance, nil
+}
+
+/**
+hbase sdk 存在问题
+*/
+func (s *HBaseService) DescribeIpWhitelist(id string) (instance hbase.Instance, err error) {
```
Comment (Lexsss) [inline alicloud/service_alicloud_hbase.go]: maybe just return a whitelist

File: examples/hbase/variables.tf [hunk 2]
```diff
@@ -0,0 +1,89 @@
+// export ALICLOUD_REGION="***"
```
Comment (Lexsss) [inline examples/hbase/variables.tf]: terraform fmt

File: website/docs/d/hbase_instances.html.markdown [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+sidebar_current: "docs-alicloud-datasource-hbase-instances"
+description: |-
+    Provides a collection of HBase instances according to the specified filters.
+---
+
+# alicloud\_hbase\_instances
+
+The `alicloud_hbase_instances` data source provides a collection of HBase instances available in Alicloud account.
+Filters support regular expression for the instance name, ids or availability_zone.
+
```
Comment (Lexsss) [inline website/docs/d/hbase_instances.html.markdown]: -> **NOTE:** Available in 1.67.0+ .

File: website/docs/r/hbase_instance.html.markdown [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  Provides a HBase instance resource.
+---
+
+# alicloud\_hbase\_instance
+
+Provides a HBase instance resource supports replica set instances only. the HBase provides stable, reliable, and automatic scalable database services. 
+It offers a full range of database solutions, such as disaster recovery, backup, recovery, monitoring, and alarms.
+You can see detail product introduction [here](https://help.aliyun.com/product/49055.html?spm=5176.124785.766162.1.67273094TcUuke)
+
+-> **NOTE:**  Available in 1.37.0+
```
Comment (Lexsss) [inline website/docs/r/hbase_instance.html.markdown]: 1.67.0+ .

File: website/docs/d/hbase_instances.html.markdown [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  availability_zone = "cn-shenzhen-b"
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name_regex` - (Optional) A regex string to apply to the instance name.
+* `ids` - (Optional, Available 1.53.0+) The ids list of HBase instances
```
Comment (Lexsss) [inline website/docs/d/hbase_instances.html.markdown]: remove `Available`

File: website/docs/d/hbase_instances.html.markdown [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  * `name` - The name of the HBase instance.
+  * `zone_id` - Zone ID the instance belongs to.
+  * `vpc_id` - VPC ID the instance belongs to.
+  * `vswitch_id` - VSwitch ID the instance belongs to.
+  * `major_version` - major version with the engine
+  * `core_node_count` - same with "core_instance_quantity"
+  * `network_type` - Classic network or VPC.
+  * `core_disk_type` - cloud_ssd or cloud_efficiency
+  * `core_instance_type` - hbase.n1.medium, hbase.n1.large, hbase.n1.2xlarge and so on.
+
```
Comment (Lexsss) [inline website/docs/d/hbase_instances.html.markdown]: please remove blank lines

File: website/docs/r/hbase_instance.html.markdown [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+}
+```
+
+this is a example for class netType instance. you can find more detail with the examples/hbase dir.
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `zone_id` - (Optional, ForceNew) The Zone to launch the DB instance. it supports multiple zone.
```
Comment (Lexsss) [inline website/docs/r/hbase_instance.html.markdown]: please sync with schema, omitted below.

File: website/docs/r/hbase_instance.html.markdown [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `id` - The ID of the HBase.
+
+### Timeouts
+
+-> **NOTE:** Available in 1.53.0+.
```
Comment (Lexsss) [inline website/docs/r/hbase_instance.html.markdown]: remove it

File: alicloud/resource_alicloud_hbase_instance.go [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"zone_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"engine": {
```
[the remark this answers is not in the document]
Comment (joyken) [reply alicloud/resource_alicloud_hbase_instance.go]: yes, hbase, hbaseue, serverspark, spark, bds.  there are all belong HBase

File: website/docs/r/hbase_instance.html.markdown (hunk 4 above)
Comment (joyken) [reply website/docs/r/hbase_instance.html.markdown]: ok, fix next commit

File: website/docs/d/hbase_instances.html.markdown (hunk 5 above)
Comment (joyken) [reply website/docs/d/hbase_instances.html.markdown]: ok,

File: website/docs/d/hbase_instances.html.markdown (hunk 6 above)
Comment (joyken) [reply website/docs/d/hbase_instances.html.markdown]: done

File: website/docs/r/hbase_instance.html.markdown (hunk 8 above)
Comment (joyken) [reply website/docs/r/hbase_instance.html.markdown]: done.

File: alicloud/data_source_alicloud_hbase_instances.go [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+			},
+			"output_file": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"names": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
```
[the remark this answers is not in the document]
Comment (joyken) [reply alicloud/data_source_alicloud_hbase_instances.go]: no finished. developing

File: alicloud/data_source_alicloud_hbase_instances_test.go [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+var checkFastInfo = dataSourceAttr{
+	resourceId:   "data.alicloud_hbase_instances.default",
+	existMapFunc: exisFastMapFunc,
+	fakeMapFunc:  fakeMapFunc,
+}
+
+// for fast test
+func TestAccAlicloudHBaseInstancesDataSourceFast(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (joyken) [reply alicloud/data_source_alicloud_hbase_instances_test.go]: yes, will remove this case next commit.

File: alicloud/data_source_alicloud_hbase_instances_test.go [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+
+var exisFastMapFunc = func(rand int) map[string]string {
+	return map[string]string{
+		"instances.#": CHECKSET,
+		"ids.#":       CHECKSET,
+		"names.#":     CHECKSET,
+	}
+}
+
+var exisMapFunc = func(rand int) map[string]string {
```
[the remark this answers is not in the document]
Comment (joyken) [reply alicloud/data_source_alicloud_hbase_instances_test.go]: ok

File: alicloud/resource_alicloud_hbase_instance.go [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Update: schema.DefaultTimeout(30 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
+		},
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"region_id": {
```
[the remark this answers is not in the document]
Comment (joyken) [reply alicloud/resource_alicloud_hbase_instance.go]: ok, can get message from zone_id

File: alicloud/resource_alicloud_hbase_instance.go [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"master_instance_type": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"core_instance_type": {
+				Type:     schema.TypeString,
+				Required: true,
```
[the remark this answers is not in the document]
Comment (joyken) [reply alicloud/resource_alicloud_hbase_instance.go]: this field can not list all, it maybe change sametimes.

File: alicloud/resource_alicloud_hbase_instance.go [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Optional: true,
+				Computed: true,
+			},
+			"status": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			// hbase 后续的参数，增加参数校验，
```
[the remark this answers is not in the document]
Comment (joyken) [reply alicloud/resource_alicloud_hbase_instance.go]: ok

File: alicloud/resource_alicloud_hbase_instance.go [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+
+	request.SecurityIPList = LOCAL_HOST_IP
+	if len(d.Get("security_ip_list").(*schema.Set).List()) > 0 {
+		request.SecurityIPList = strings.Join(expandStringList(d.Get("security_ip_list").(*schema.Set).List())[:], COMMA_SEPARATED)
+	}
+	return request, nil
+}
+
+func resourceAlicloudHBaseInstanceCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Println("[DEBUG] hbase-log id=", d)
```
[the remark this answers is not in the document]
Comment (joyken) [reply alicloud/resource_alicloud_hbase_instance.go]: done.

File: alicloud/resource_alicloud_hbase_instance.go [hunk 17]
```diff
@@ -0,0 +1,309 @@
...
+	fmt.Println(instance)
+	if err != nil {
+		if NotFoundError(err) {
+			d.SetId("")
+			return nil
+		}
+		return WrapError(err)
+	}
+
+	_ = d.Set("name", instance.InstanceName)
```
[the remark this answers is not in the document]
Comment (joyken) [reply alicloud/resource_alicloud_hbase_instance.go]: done

File: alicloud/resource_alicloud_hbase_instance.go [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+	_ = d.Set("core_instance_type", instance.CoreInstanceType)
+	_ = d.Set("core_instance_quantity", instance.CoreNodeCount)
+	_ = d.Set("core_disk_size", instance.CoreDiskSize)
+	_ = d.Set("core_disk_type", instance.CoreDiskType)
+	_ = d.Set("pay_type", instance.PayType)
+	_ = d.Set("status", instance.Status)
+	_ = d.Set("net_type", instance.NetworkType)
+	_ = d.Set("vpc_id", instance.VpcId)
+	_ = d.Set("vswitch_id", instance.VswitchId)
+	log.Println("[DEBUG] hbase-log result d=", d)
```
[the remark this answers is not in the document]
Comment (joyken) [reply alicloud/resource_alicloud_hbase_instance.go]: done

File: alicloud/service_alicloud_hbase.go [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+func (s *HBaseService) NotFoundHBaseInstance(err error) bool {
+	if NotFoundError(err) || IsExceptedErrors(err, []string{InvalidHBaseInstanceIdNotFound, InvalidHBaseNameNotFound}) {
+		return true
+	}
+	return false
+}
+
+func (s *HBaseService) DescribeHBaseInstance(id string) (instance hbase.Instance, err error) {
```
[the remark this answers is not in the document]
Comment (joyken) [reply alicloud/service_alicloud_hbase.go]: good idea

File: website/docs/d/hbase_instances.html.markdown (hunk 3 above)
Comment (joyken) [reply website/docs/d/hbase_instances.html.markdown]: done

File: website/docs/r/hbase_instance.html.markdown (hunk 7 above)
Comment (joyken) [reply website/docs/r/hbase_instance.html.markdown]: done

**Pull request closed without merging**